pinot-spi/src/main/java/org/apache/pinot/spi/environmentprovider/PinotEnvironmentProvider.java

```diff
@@ -0,0 +1,40 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.environmentprovider;
+
+import java.util.Map;
+import org.apache.pinot.spi.env.PinotConfiguration;
+
+/**
+ *  Environment Provider interface implemented by different cloud providers to customize
+ *  the base pinot configuration to add environment variables & instance specific configuration
+ */
+public interface PinotEnvironmentProvider {
+
+  /**
+   * Initializes the configurations specific to an environment provider.
+   */
+  void init(PinotConfiguration pinotConfiguration);
+
+  /**
+   * Customize base pinot configuration to add environment variables & instance specific configuration
+   * @return custom pinot configuration map
+   */
+  Map<String, String> getEnvironment();
```

Comment:
So, every cloud provider can provide any random map, and we add the map to zookeeper? But then the only thing that Pinot can understand and work with, is failure domain. Can you give me an example of what Pinot could do if a cloud provider adds other keys to the map?

Reply:
The plan is to add all the environment properties under mapFields. If u see the getEnvironment() method returns a Map. For now we are only adding FD information, but other cloud providers can add other environment properties. At HelixServerStarter layer we will update instanceConfig with the entire map returned by getEnvironment() call. 

Currently, the only thing pinot can derive sense out of is the failure domain information. Pinot can use FD information to assign server to server pools to facilitate segment & replica group assignments.